pubsub/iterator.go

```diff
@@ -232,17 +259,32 @@ func (it *messageIterator) receive(maxToPull int32) ([]*Message, error) {
 	// We received some messages. Remember them so we can keep them alive. Also,
 	// do a receipt mod-ack when streaming.
 	maxExt := time.Now().Add(it.po.maxExtension)
-	ackIDs := map[string]bool{}
+	ackIDs := map[string]*AckResult{}
 	it.mu.Lock()
 	for _, m := range msgs {
 		ackID := msgAckID(m)
 		addRecv(m.ID, ackID, now)
 		it.keepAliveDeadlines[ackID] = maxExt
 		// Don't change the mod-ack if the message is going to be nacked. This is
 		// possible if there are retries.
-		if !it.pendingNacks[ackID] {
-			ackIDs[ackID] = true
+		if _, ok := it.pendingNacks[ackID]; !ok {
+			// Don't use the message's AckResult here since these are only for receipt modacks.
+			// ModAckResults are transparent to the user anyway so these can automatically succeed.
+			// We can't use an empty AckResult here either since SetAckResult will try to
+			// close the channel without checking if it exists.
+			ackIDs[ackID] = newSuccessAckResult()
+		}
+		// If exactly once is enabled, keep track of all pending AckResults
+		// so we can cleanly close them all at shutdown.
+		it.eoMu.Lock()
```

Comment:
Any concern about performance impact to non exactly once delivery subs due to these locks, here and below?

can the critical section be reduced if we read the value of `it.enableExactlyOnceDelivery` into a local variable inside a reader lock, and then execute the rest of the logic using that variable? Would that impact the logical correctness?

Reply:
Yeah the lock acquisition should've been a `RLock() + RUnlock()` here and outside of the loop. I made a similar optimization (moving the mutex locking) in other places as well. The only place where the mutex is locked for writing is when receiving on the stream.